src/betree.h

```diff
@@ -0,0 +1,139 @@
+#ifndef BETREE_H__
+#define BETREE_H__
+
+#include <stdbool.h>
+#include <stdlib.h>
+
+struct pred {
+    unsigned int variable_id;
+    int value;
+};
+
+struct sub {
+    unsigned int id;
+    unsigned int variable_id_count;
+    // TODO Make const
+    unsigned int* variable_ids;
+    const struct ast_node *expr;
+};
+
+struct event {
+    unsigned int pred_count;
+    // TODO Make const
+    struct pred** preds;
```

Comment:
You can skip an extra allocation and lots of point headaches here by using the following trick:
```
struct event { size_t len; struct pred* preds[]; };
struct event *ev = calloc(1, sizeof(*ev) + len * sizeof(ev->preds[0]));
ev->len = len;

// ...

ev->preds[i];
```

While this required that the size of the array not change after it's been created (I still have to confirm this in the code) but it generally simplifies the and reduces the amount of frees you have to remember to do :)

Reply:
At first I made them all static with a size indeed but I ended up using malloc/free in the end. I run valgrind often enough and there shouldn't be any leaks. I guess I just didn't like how dumb vscode is and woould try to show the whole array. inb4 >vscode